docs/tutorial/frontend-tut.md

```diff
@@ -1,94 +1,117 @@
+In the [Calyx tutorial][calyx-tut] you wrote Calyx code by hand.
+This is (probably) a good way to build character, but it's no way to live.
+Indeed, Calyx was _designed_ to be a compiler IL, and not a human-facing language.
 
-```
-{{#include ../../frontends/mrxl/test/dot.mrxl}}
-```
+In this tutorial, we're going to learn all about this by building a DSL-to-hardware compiler for a toy language that we wish to accelerate.
+We will compile the DSL to Calyx, and then let Calyx take us to hardware.
+
+
+# MrXL Overview
 
-We define the interface of program by specifying `input` and `output` arrays.
-Input arrays have their values populated by an external harness while the output arrays must be computed using the program.
+Meet MrXL, our toy DSL.
+MrXL lets you define arrays and registers and then perform `map` and `reduce` operations.
 
-A `map` expression iterates over multiple arrays of the same element and produces a new vector using the function provided in the body.
-In the above example, the `map` expression multiplies the values of `avec` and `bvec`.
-`map 1` states that the operation has a *parallelism factor* of 1 which means that the loop iterations are performed sequentially.
 
-`reduce` expressions walk over memories and accumulate a result into a register.
-In the above code snippet, we add together all the elements of `prodvec` and place them in a register named `dot`.
-Since the `reduce` parallelism factor is also 1, the reduction is performed sequentially.
+## Example: sum of squares
+
+Here's a MrXL program that squares and then sums the values of an input array:
+```
+{{#include ../../frontends/mrxl/test/sos.mrxl}}
+```
+
+This short program shows off all of MrXL's features, so let's pick it apart line by line:
+1. We specify an array, `avec`, which will have four integers. The `input` keyword means that an external harness will populate the array.
+2. We specify `sos`, a register. The `output` keyword means that we will populate `sos` in our program.
+3. The `map` operation gets the values of `avec` and raises each to the second power. We stash the result in a new array, `squares`. The number `1` denotes a *parallelism factor* of 1, meaning that the operation is performed sequentially. We will improve this shortly.
+4. The `reduce` operation walks over `squares` and accumulates the result into a register. The parallelism factor is again `1`.
 
 
-## Run a MrXL Program
+## Running our example
 
-Once we have [installed the mrxl command line tool][mrxldocs], we can run MrXL programs using [`fud`][fud].
+Let's run this program.
 
-To provide MrXL program with input values, we use fud's [JSON][json]-based [data format][fud-data].
-Let's try to run this program, which has a parallelism factor of two:
+To begin, [install the MrXL command line tool][mrxldocs-install].
+(TK: I'd kinda like to inline those instructions here, so that there is less jumping around between files.
```

Comment:
Sounds good! I think you can also possibly use a `<details>` tag and make it dropdownable? The [dahlia-evaluation README](https://github.com/cucapra/dahlia-evaluation/blob/7e1cb8f1ad261dc091b9ffd8adec41f042f5ec96/README.md?plain=1#L450)

Reply:
They sorta need all the text, so I'm not sure what the "summary" of the `details` tag would be... I've just put it in as a list, happy to play with it further.